### PR TITLE
[RFC] Add new database object / fieldset - Stage 0

### PR DESCRIPTION
**Pull request by mbrancato**
Add new database object / fieldset to cover multiple database use cases.

Related to #514 
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/ecs/blob/master/CONTRIBUTING.md)?
- For proposing substantial changes or additions to the schema, have you reviewed the [RFC process](https://github.com/elastic/ecs/blob/master/rfcs/README.md)?
- If submitting code/script changes, have you verified all tests pass locally using `make test`?
- If submitting schema/fields updates, have you generated new artifacts by running `make` and committed those changes?
- Is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- Have you added an entry to the [CHANGELOG.next.md](https://github.com/elastic/ecs/blob/master/CHANGELOG.next.md)?

### CODE REVIEW

**Comment by djptek** [comment]
Hi @mbrancato the [PR 514](https://github.com/elastic/ecs/issues/514) referenced mentions the APM `span.db.*`fields which currently include:

```
span.db.statement
span.db.type
span.db.instance
span.db.link
span.db.user
```

They are not currently in ECS, however, the APM functionality in Elastic Stack already provides a host of plug-and-play tools for capturing and analyzing database events as transactions and spans, here is a [blog showing how to monitor mysql transactions in a java app](https://www.elastic.co/blog/monitoring-java-applications-and-getting-started-with-the-elastic-apm-java-agent)

Do you think the [fields under span.db.*](https://www.elastic.co/guide/en/apm/server/7.13/exported-fields-apm-span.html) might fulfill your requirement if they were in ECS either as custom or as part of the standard?

**Comment by mbrancato** [comment]
Hi @djptek - I think I first ran into this looking at Beats and then stumbled into that GH issue. As in that GitHub issue, the problem is there are already several tools doing different schemas to capture database information. For my specific use-case the top-level `span` is more confusing as they are not tracing spans. But surely, a reusable database field set could be placed at `span.db` or anywhere else.

I do think there would still be postgres, mysql, and other specific schemas, but having a common place to record this information makes querying the data easier.

At a minimum, I think there should be a single field set that can identify:
- DB type
- DB name(s) - this could be an array
- data container type (table, index, collection, etc)
- data container(s) - this should be an array
- query (raw query)
- fields (if any)
- operation (main operation name like SELECT, $match, _bulk)

I'm thinking nested statements should be split up if needed into multiple events.

My use case here is security related. it allows building rules around possibly queries to inserts into specific databases or tables without needing technology-specific fields. To me, having a top-level DB field set is just like having `source` or `destination` fieldsets. Trying to capture data under DB-specific technologies is like trying to have specific link-layer fieldsets like Ethernet, Frame Relay, DSL, etc.

**Comment by djptek** [comment]
>> I'm thinking nested statements should be split up if needed into multiple events.

That would work really well with the APM tooling - even though the use case is Security it may still be useful to navigate up and down a stack of nested queries

re: `span.db.statement` or similar, you could get the best of both worlds 

- with slightly increased storage cost by copying selected data from e.g. field(s) such as you are proposing
- or with slightly increased complexity by aliasing the field(s)

File: rfcs/text/0000-database-object.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+## References
+
+https://github.com/elastic/ecs/issues/514
+<!-- Insert any links appropriate to this RFC in this section. -->
+
+### RFC Pull Requests
+
+<!-- An RFC should link to the PRs for each of it stage advancements. -->
+
+* Stage 0: https://github.com/elastic/ecs/pull/NNN
```
Comment (ebeahan) [inline rfcs/text/0000-database-object.md:103]:
```suggestion
* Stage 0: https://github.com/elastic/ecs/pull/1428
```

File: rfcs/text/0000-database-object.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# 0000: Database Object
```
Comment (ebeahan) [inline rfcs/text/0000-database-object.md:1]:
```suggestion
# 0000: Database Field Set
```

Any objection (no pun intended 😂 ) to rephrasing the title?

**Review by ebeahan — APPROVED** [review]
Thanks, @mbrancato, for putting this together!

For a stage 0 RFC PR, the [criteria](https://elastic.github.io/ecs/stages.html) is kept pretty simple: `The premise of these changes is not obviously useless or inappropriate for ECS`. A database-focused field set has been discussed before, and I do think the concept would be an appropriate addition to ECS.

I agree, though, that as this RFC advances we'll want to consider how any proposed fields relate to the existing `span.db.` fields.

A couple of other thoughts (no need to address/comment now - we can discuss further as the proposal advances):

* `db.*` or `database.*`? ECS tries to avoid abbreviations when possible, but it also makes exceptions for well-known concepts like `ip`, `geo`, `os`. I think `db.*` is definitely a well-known shorthand, but the full `database.*` is also short. So I'm on the fence 😄 
* I'm not clear what type of use cases would need to reuse a `db.*` field set. Again, something we can discuss in a later stage PR.

**Comment by mbrancato** [comment]
Thanks. I did have one other use case I identified. That is being able to log functions called. It would work along side fields in my current mockup.

- fields (if any)
- functions (if any)
  - fields (if any, for value-only functions)

**Comment by ebeahan** [comment]
Thanks, @mbrancato.

@djptek, did you have anything you wanted to continue discussing before we move this proposal to stage 0?

**Comment by djptek** [comment]
> Thanks, @mbrancato.
> 
> @djptek, did you have anything you wanted to continue discussing before we move this proposal to stage 0?

LGTM!

**Comment by djptek** [comment]
@mbrancato please use RFC 0026

**Comment by ebeahan** [comment]
Thanks, all. I'll go ahead and set the RFC number, set the date, and then merge.

**Pull request merged**